web/pdf_viewer.js

```diff
@@ -2259,13 +2293,54 @@ class PDFViewer {
     if (!this.pdfDocument) {
       return;
     }
-    this.#annotationEditorMode = mode;
-    this.eventBus.dispatch("annotationeditormodechanged", {
-      source: this,
-      mode,
-    });
 
-    this.#annotationEditorUIManager.updateMode(mode, editId, isFromKeyboard);
+    const { eventBus } = this;
+    const updater = () => {
+      if (this.#onPageRenderedCallback) {
+        eventBus._off("pagerendered", this.#onPageRenderedCallback);
+        this.#onPageRenderedCallback = null;
+      }
+      if (this.#switchAnnotationEditorModeTimeoutId !== null) {
+        clearTimeout(this.#switchAnnotationEditorModeTimeoutId);
+        this.#switchAnnotationEditorModeTimeoutId = null;
+      }
+      this.#annotationEditorMode = mode;
+      eventBus.dispatch("annotationeditormodechanged", {
+        source: this,
+        mode,
+      });
+      this.#annotationEditorUIManager.updateMode(mode, editId, isFromKeyboard);
+    };
+
+    if (
+      mode === AnnotationEditorType.NONE ||
+      this.#annotationEditorMode === AnnotationEditorType.NONE
+    ) {
+      const isEditing = mode !== AnnotationEditorType.NONE;
+      if (!isEditing) {
+        this.pdfDocument.annotationStorage.resetModifiedIds();
+      }
+      for (const pageView of this._pages) {
+        pageView.toggleEditingMode(isEditing);
+      }
+      const idsToRefresh = this.#switchToEditAnnotationMode();
+      if (isEditing && editId && idsToRefresh) {
+        // We're editing an existing annotation so we must switch to editing
+        // mode when the rendering is done.
+        const { signal } = this.#eventAbortController;
+        this.#onPageRenderedCallback = ({ pageNumber }) => {
+          idsToRefresh.delete(pageNumber);
+          if (idsToRefresh.size === 0) {
+            eventBus._off("pagerendered", this.#onPageRenderedCallback);
+            this.#onPageRenderedCallback = null;
+            this.#switchAnnotationEditorModeTimeoutId = setTimeout(updater, 0);
+          }
+        };
+        eventBus._on("pagerendered", this.#onPageRenderedCallback, { signal });
+        return;
+      }
```

Comment:
Given that looking up the currently visible pages isn't "free" in larger/longer documents, how about trying to avoid that in *all cases* where that's possible?
```suggestion
      if (isEditing && editId) {
        const idsToRefresh = this.#switchToEditAnnotationMode();
        if (idsToRefresh) {
          // We're editing an existing annotation so we must switch to editing
          // mode when the rendering is done.
          const { signal } = this.#eventAbortController;
          this.#onPageRenderedCallback = ({ pageNumber }) => {
            idsToRefresh.delete(pageNumber);
            if (idsToRefresh.size === 0) {
              eventBus._off("pagerendered", this.#onPageRenderedCallback);
              this.#onPageRenderedCallback = null;
              this.#switchAnnotationEditorModeTimeoutId = setTimeout(updater, 0);
            }
          };
          eventBus._on("pagerendered", this.#onPageRenderedCallback, { signal });
          return;
        }
      }
```